### PR TITLE
[fix][broker] Avoid block markDeletePosition forward when skip lost entries

### PR DESCRIPTION
**Pull request by hanmz**
### Motivation

The configuration autoSkipNonRecoverableData is designed to turn this feature on if we can accept partial data loss. When some entries is lost, the broker will still work. `cursor.readPosition` will move to next position for skip these positions. But now we have this problem: `cursor.readPosition` can move to next positition, but the `cursor.markDeletePosition ` can not forward.

for example:
1. topic-1 has some entries:  [1:1],[1:2],[1:3],[1:4],[1:5],[1:6],[1:7],[1:8]
2. entry [1,4] and [1,5] lost
3. cursor read entry from ledger, then send message [1:1],[1:2],[1:3],[1:6],[1:7] to consumer. (Current: readPosition=[1:8], markDeletePosition=[1:1])
4. consumer individual ack [1:1],[1:2],[1:3],[1:6],[1:7]  (Current: readPosition=[1:8], markDeletePosition=[1:3])
5. We can't delete the entry [1:4] and [1:5] because these lost. So markDeletePosition will stay at [1:3] and can not move forward to [1:8]

The root cause is: When entries lost, lost entries not be delete from cursor.

### Modifications

When skip lost entries, delete these entries from cursor.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository
PR in forked repository: https://github.com/hanmz/pulsar/pull/4

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -2791,6 +2791,54 @@ public void deleteFailed(ManagedLedgerException ex, Object ctx) {
...
 
+    /**
+     * Manually acknowledge all entries from startPosition to endPosition.
+     * - Since this is an uncommon event, we focus on maintainability. So we do not modify
+     *   {@link #individualDeletedMessages} and {@link #batchDeletedIndexes}, but call
+     *   {@link #asyncDelete(Position, AsyncCallbacks.DeleteCallback, Object)}.
+     * - This method is valid regardless of the consumer ACK type.
+     * - If there is a consumer ack request after this event, it will also work.
+     */
+    public void skipNonRecoverableEntries(Position startPosition, Position endPosition){
```
Comment (Jason918) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:2802]: It's better to add a unit test for this method.

**Review by Jason918 — APPROVED** [review]
overall LGTM.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpReadEntry.java [hunk 2]
```diff
@@ -135,6 +135,8 @@ public void readEntriesFailed(ManagedLedgerException exception, Object ctx) {
             updateReadPosition(nexReadPosition);
             if (lostLedger != null) {
                 cursor.getManagedLedger().skipNonRecoverableLedger(lostLedger);
+            } else {
+                cursor.skipNonRecoverableEntries(readPosition, nexReadPosition);
```
Comment (poorbarcode) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpReadEntry.java:139]: Could you add a test for this scenario?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java (hunk 1 above)
Comment (hanmz) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:2802]: Okay, I've added a unit test

**Review by Jason918 — APPROVED** [review]
LGTM

**Review by lhotari — APPROVED** [review]

**Comment by hanmz** [comment]
If you have time, please help me take a look, thank you very much. @Technoboy- @poorbarcode

**Pull request merged**